transport-classes-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollChannel.java

```diff
@@ -167,48 +167,61 @@ public ChannelMetadata metadata() {
-        inputClosedSeenErrorOnRead = true;
-        try {
-            ChannelPromise promise = connectPromise;
-            if (promise != null) {
-                // Use tryFailure() instead of setFailure() to avoid the race against cancel().
-                promise.tryFailure(new ClosedChannelException());
-                connectPromise = null;
-            }
+    protected void doClose(ChannelPromise promise) {
+        ChannelPromise connectPromise = this.connectPromise;
+        if (connectPromise != null) {
+            // Use tryFailure() instead of setFailure() to avoid the race against cancel().
+            promise.tryFailure(new ClosedChannelException());
+            this.connectPromise = null;
+        }
 
-            Future<?> future = connectTimeoutFuture;
-            if (future != null) {
-                future.cancel(false);
-                connectTimeoutFuture = null;
-            }
+        Future<?> future = connectTimeoutFuture;
+        if (future != null) {
+            future.cancel(false);
+            connectTimeoutFuture = null;
+        }
 
-            if (isRegistered()) {
-                // Need to check if we are on the EventLoop as doClose() may be triggered by the GlobalEventExecutor
-                // if SO_LINGER is used.
-                //
-                // See https://github.com/netty/netty/issues/7159
-                EventLoop loop = executor();
-                if (loop.inEventLoop()) {
-                    doDeregister();
-                } else {
-                    loop.execute(new Runnable() {
-                        @Override
-                        public void run() {
-                            try {
-                                doDeregister();
-                            } catch (Throwable cause) {
-                                pipeline().fireExceptionCaught(cause);
-                            }
-                        }
-                    });
+        if (isRegistered()) {
+            // Need to check if we are on the EventLoop as doClose() may be triggered by the GlobalEventExecutor
+            // if SO_LINGER is used.
+            //
+            // See https://github.com/netty/netty/issues/7159
+            EventLoop loop = executor();
+            ChannelPromise deregisterPromise = newPromise();
+            deregisterPromise.addListener(f -> {
+                active = false;
+                // Even if we allow half closed sockets we should give up on reading. Otherwise we may allow a read
+                // attempt on a socket which has not even been connected yet. This has been observed to block during
+                // unit tests.
+                inputClosedSeenErrorOnRead = true;
+                try {
+                    socket.close();
+                } catch (Throwable cause) {
+                    promise.setFailure(cause);
                 }
+                promise.setSuccess();
```

Comment:
This should be moved up after `socket.close()` because if that fails, we'll throw an `IllegalStateException` here.

Reply:
just added return statements after the `setFailure` as otherwise it would be possible for `promise.setSuccess()` to throw and be catches by `setFailure` which would throw as well